src/librepgp/stream-parse.h

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (c) 2017, [Ribose Inc](https://www.ribose.com).
+ * Copyright (c) 2009 The NetBSD Foundation, Inc.
+ * All rights reserved.
+ *
+ * This code is originally derived from software contributed to
+ * The NetBSD Foundation by Alistair Crooks (agc@netbsd.org), and
+ * carried further by Ribose Inc (https://www.ribose.com).
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef STREAM_PARSE_H_
+#define STREAM_PARSE_H_
+
+#include <stdint.h>
+#include <stdbool.h>
+#include <sys/types.h>
+#include "errors.h"
+
+#define PGP_INPUT_CACHE_SIZE 32768
+
+typedef enum {
+    PGP_SOURCE_FILE,
+    PGP_SOURCE_MEMORY,
+    PGP_SOURCE_STDIN,
+    PGP_SOURCE_PACKET,
+    PGP_SOURCE_PARLEN_PACKET,
+    PGP_SOURCE_LITERAL,
+    PGP_SOURCE_ENCRYPTED,
+    PGP_SOURCE_ARMOURED,
+    PGP_SOURCE_CLEARTEXT
+} pgp_source_type_t;
+
+typedef struct pgp_source_t pgp_source_t;
+typedef ssize_t pgp_source_read_func_t(pgp_source_t *src, void *buf, size_t len);
+typedef void pgp_source_close_func_t(pgp_source_t *src);
+
+/* statically preallocated cache for sources. Not used for input filters */
+typedef struct pgp_source_cache_t {
+    uint8_t buf[PGP_INPUT_CACHE_SIZE];
+    unsigned pos;
+    unsigned len;
+} pgp_source_cache_t;
+
+typedef struct pgp_source_t {
+    pgp_source_read_func_t  *readfunc;
+    pgp_source_close_func_t *closefunc;
```

Comment:
Minor detail but instead of `readfunc`/`closefunc` for field names, I think `read`/`close` or `reader`/`closer` would be fine (like how we currently have `pgp_writer_func_t *writer`). It's obvious from context at the call sites that it's a function, so it seems overly verbose maybe.

Reply:
Agree on this, will rename.